src/main/factsManager.js

```diff
@@ -0,0 +1,235 @@
+ * the user has been keyed in). Relies on fact providers to function properly.
+ * Copyright 2017 Raising the Floor - International
+ *
+ * Licensed under the New BSD license. You may not use this file except in
+ * compliance with this License.
+ * The research leading to these results has received funding from the European Union's
+ * Seventh Framework Programme (FP7/2007-2013) under grant agreement no. 289016.
+ * You may obtain a copy of the License at
+ * https://github.com/GPII/universal/blob/master/LICENSE.txt
+ */
+"use strict";
+
+var fluid = require("infusion");
+
+var gpii = fluid.registerNamespace("gpii");
+
+require("./utils.js");
+
+/**
+ * Manages and disseminates facts. A fact is simply represented by a name
+ * and value. For each fact there is a fact provider that collects and supplies it.
+ * Example of a fact is "keyedInBefore", which defines times since keying in in
+ * milliseconds.
+ *
+ * Once a fact's updated notification is recieved, all facts are collected from the providers
+ * and sent through the `onFactsUpdated` event.
+ */
+fluid.defaults("gpii.app.factsManager", {
+    gradeNames: ["fluid.component"],
+
+    events: {
+        /*
+         * Listened to from component users
+         * Sent with a map of facts that follow the schema:
+         * {
+         *  "factName1": factValue1,
+         *  "factName2": factValue2,
+         *  ...
+         * }
+         */
+        onFactsUpdated: null
+    },
+
+    components: {
+        /*
+         * Fact Providers
+         */
+        keyedInBeforeProvider: {
+            type: "gpii.app.factsManager.keyedInBeforeProvider",
+            options: {
+                listeners: {
+                    onFactUpdated: "{factsManager}.emitFacts"
+                }
+            }
+        }
+    },
+
+    invokers: {
+        getFacts: {
+            // Simple iteration over all facts providers
+            funcName: "gpii.app.factsManager.getFacts",
+            args: [
+                "{that}"
+            ]
+        },
+        emitFacts: {
+            funcName: "gpii.app.factsManager.emitFacts",
+            args: ["{that}"]
+        },
+
+        resetFacts: {
+            // Simple iteration over all facts providers
+            funcName: "gpii.app.factsManager.resetFacts",
+            args: ["{that}"]
+        }
+    }
+});
+
+/**
+ * Simply sends all freshly collected facts over the responsible event.
+ * @param that {Component} The `gpii.app.factsManager` component
+ */
+gpii.app.factsManager.emitFacts = function (that) {
+    that.events.onFactsUpdated.fire(that.getFacts());
+};
+
+/**
+ * Get facts collected from all fact providers.
+ * @param that {Component} The gpii.app.factsManager component
+ * @return {Object} A map of all facts - { factName: factValue, ... },
+ * where the `factName` is defined in the corresponding provider
+ */
+gpii.app.factsManager.getFacts = function (that) {
+    var factProviders = gpii.app.getSubcomponents(that);
```

Comment:
To avoid races of the kind you are worried about, it would be clearer if this relationship was mediated by an options distribution from the factManager to the onCreate event of the fact providers, which directs them to register themselves with the factManager. The cleanest expression would just write this in a grade which is distributed onto the factProviders.